### PR TITLE
K8s

### PR DESCRIPTION
**Pull request by d33d33**

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#81.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#81.0 ✔ 
* 
  * checks ✔

**Comment by yanndegat** [comment]
hi @d33d33, @PhilippeVienne 

thanks for this PR, this is great work.

first global remark: some files should be renamed accordlingly to the resources/datasources they define.
eg: ovh/resource_ovh_cloud_kubernetes_cluster.go -> ovh/resource_ovh_cloud_project_kube.go


ovh/data_source_ovh_cloud_kubernetes_cluster.go 
ovh/data_source_ovh_cloud_kubernetes_cluster_test.go
ovh/resource_ovh_cloud_kubernetes_cluster.go
ovh/resource_ovh_cloud_kubernetes_node_pool_test.go 
ovh/resource_ovh_cloud_kubernetes_node_pool.go
ovh/resource_ovh_cloud_kubernetes_cluster_test.go

File: ovh/resource_ovh_cloud_kubernetes_cluster.go [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+	}
+
+	err = nil
+	return
+}
+
+func waitForCloudProjectKubeActive(c *ovh.Client, projectId, cloudProjectKubeId string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		r := &PublicCloudProjectKubeResponse{}
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_cluster.go:320]: should be CloudProjectKubeResponse

File: ovh/resource_ovh_cloud_kubernetes_cluster.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+			return r, "", err
+		}
+
+		return r, r.Status, nil
+	}
+}
+
+func waitForCloudProjectKubeDelete(c *ovh.Client, projectId, CloudProjectKubeId string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		r := &PublicCloudProjectKubeResponse{}
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_cluster.go:333]: CloudProjectKubeResponse

File: ovh/resource_ovh_cloud_kubernetes_cluster_test.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	"os"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+var testAccPublicCloudProjectKubeConfig = fmt.Sprintf(`
+resource "ovh_cloud_project_kube" "cluster" {
+	project_id  = "%s"
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_cluster_test.go:14]: should be service_name

File: ovh/resource_ovh_cloud_kubernetes_cluster_test.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  	name = "my cluster for acceptance tests"
+	region = "GRA5"
+	version = "1.15"
+}
+`, os.Getenv("OVH_PUBLIC_CLOUD"))
+
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_cluster_test.go:20]: a sweeper function has to be implemented

example here
https://github.com/ovh/terraform-provider-ovh/blob/master/ovh/resource_ovh_iploadbalancing_tcp_farm_test.go#L56

File: ovh/resource_ovh_cloud_kubernetes_node_pool_test.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+resource "ovh_cloud_project_kube_node_pool" "pool" {
+	project_id  = "%s"
+	cluster_id = data.ovh_cloud_project_kube.cluster.id
+  	name = "acceptance-node"
+	flavor = "b2-7"
+	desired_size = 1
+	min_size = 0
+	max_size = 1
+}
+`, os.Getenv("OVH_PUBLIC_CLOUD"), os.Getenv("OVH_KUBERNETES_CLUSTER_NAME"), os.Getenv("OVH_PUBLIC_CLOUD"))
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool_test.go:26]: should be OVH_CLOUD_PROJECT_SERVICE_TEST and OVH_CLOUD_PROJECT_KUBE_NAME_TEST

File: ovh/resource_ovh_cloud_kubernetes_node_pool_test.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	"os"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+var testAccPublicCloudProjectKubeNodePoolConfig = fmt.Sprintf(`
+data "ovh_cloud_project_kube" "cluster" {
+	project_id  = "%s"
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool_test.go:14]: shloud be service_name

File: ovh/types_cloud.go [hunk 7]
```diff
@@ -176,3 +176,75 @@ type CloudServiceStatusResponse struct {
 func (s *CloudServiceStatusResponse) String() string {
 	return fmt.Sprintf("%s: %s", s.Name, s.Status)
 }
+
+type PublicCloudProjectKubeResponse struct {
```
Comment (yanndegat) [inline ovh/types_cloud.go:180]: shloud be CloudProjectKubeResponse

File: ovh/types_cloud.go [hunk 8]
```diff
@@ -176,3 +176,75 @@ type CloudServiceStatusResponse struct {
...
+	Flavor     string `json:"flavor"`
+	Status     string `json:"status"`
+	Version    string `json:"version"`
+}
+
+func (n *CloudProjectKubeNodeResponse) String() string {
+	return fmt.Sprintf("%s(%s): %s", n.Name, n.Id, n.Status)
+}
+
+type CloudProjectKubeNodePoolCreationRequest struct {
```
Comment (yanndegat) [inline ovh/types_cloud.go:217]: should be CloudProjectKubeNodePoolCreateOpts

File: ovh/types_cloud.go [hunk 9]
```diff
@@ -176,3 +176,75 @@ type CloudServiceStatusResponse struct {
...
+	DesiredNodes  int    `json:"desiredNodes"`
+	MaxNodes      int    `json:"maxNodes"`
+	MinNodes      int    `json:"minNodes"`
+	FlavorName    string `json:"flavorName"`
+	Name          string `json:"name"`
+	MonthlyBilled bool   `json:"monthlyBilled"`
+	AntiAffinity  bool   `json:"antiAffinity"`
+}
+
+type CloudProjectKubeNodePoolUpdateRequest struct {
```
Comment (yanndegat) [inline ovh/types_cloud.go:227]: shloud be CloudProjectKubeNodePoolUpdateOpts

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+
+		Importer: &schema.ResourceImporter{
+			StateContext: func(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				err := resourceCloudProjectKubeNodePoolRead(d, meta)
+				return []*schema.ResourceData{d}, err
+			},
+		},
+
+		Schema: map[string]*schema.Schema{
+			"project_id": {
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:31]: should be service_name

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				DefaultFunc: schema.EnvDefaultFunc("OVH_PROJECT_ID", nil),
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:35]: should be OVH_CLOUD_PROJECT_SERVICE

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceCloudProjectKubeNodePoolUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) (diags diag.Diagnostics) {
+	config := meta.(*Config)
+
+	projectId := d.Get("project_id").(string)
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:87]: should be service_name

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 13]
```diff
@@ -0,0 +1,405 @@
...
+	d.Partial(false)
+
+	return nil
+
+}
+
+func resourceCloudProjectKubeNodePoolCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) (diags diag.Diagnostics) {
+	config := meta.(*Config)
+
+	projectId := d.Get("project_id").(string)
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:161]: should be service_name

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+
+	d.Partial(false)
+
+	return nil
+}
+
+func resourceCloudProjectKubeNodePoolRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	projectId := d.Get("project_id").(string)
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:264]: should be service_name

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+	d.Partial(false)
+
+	log.Printf("[DEBUG] Read nodepool: %+v", r)
+	return nil
+}
+
+func resourceCloudProjectKubeNodePoolDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) (diags diag.Diagnostics) {
+	config := meta.(*Config)
+
+	projectId := d.Get("project_id").(string)
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:293]: should be service_name

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 16]
```diff
@@ -0,0 +1,405 @@
...
+}
+
+func readCloudProjectKubeNodePool(projectId string, config *Config, d *schema.ResourceData, cluster *CloudProjectKubeNodePoolResponse) (err error) {
+	_ = d.Set("name", cluster.Name)
+	_ = d.Set("flavor", cluster.Flavor)
+	_ = d.Set("desired_nodes", cluster.DesiredNodes)
+	_ = d.Set("max_nodes", cluster.MaxNodes)
+	_ = d.Set("min_nodes", cluster.MinNodes)
+	_ = d.Set("monthly_billed", cluster.MonthlyBilled)
+	_ = d.Set("status", cluster.Status)
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:354]: remove _=

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 17]
```diff
@@ -0,0 +1,405 @@
...
+	clusterId := d.Get("cluster_id").(string)
+	poolId := d.Id()
+
+	params := &CloudProjectKubeNodePoolUpdateRequest{
+		DesiredNodes: d.Get("desired_nodes").(int),
+		MaxNodes:     d.Get("max_nodes").(int),
+		MinNodes:     d.Get("min_nodes").(int),
+	}
+
+	d.Partial(true)
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:97]: to be removed

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 18]
```diff
@@ -0,0 +1,405 @@
...
+		diags = append(diags, diag.Diagnostic{
+			Severity:      diag.Error,
+			Summary:       "Failed to read nodepool",
+			Detail:        err.Error(),
+			AttributePath: nil,
+		})
+		return
+	}
+
+	d.Partial(false)
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:152]: to be removed

Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:152]: return resourceCloudProjectKubeNodePoolRead(d, meta)

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 19]
```diff
@@ -0,0 +1,405 @@
...
+		FlavorName:    d.Get("flavor").(string),
+		DesiredNodes:  d.Get("desired_nodes").(int),
+		MaxNodes:      d.Get("max_nodes").(int),
+		MinNodes:      d.Get("min_nodes").(int),
+		MonthlyBilled: d.Get("monthly_billed").(bool),
+	}
+
+	r := &CloudProjectKubeNodePoolResponse{}
+
+	d.Partial(true)
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:175]: to be removed

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+	log.Printf("[DEBUG] Will create nodepool: %+v", params)
+
+	err := config.OVHClient.Post(endpoint, params, r)
+	if err != nil {
+		diags = append(diags, diag.Diagnostic{
+			Severity:      diag.Error,
+			Summary:       "Failed to create nodepool",
+			Detail:        err.Error(),
+			AttributePath: nil,
+		})
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:188]: we dont do this elsewhere, return err instead

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+	_, err = bugFixWait.WaitForStateContext(ctx)
+	if err != nil {
+		diags = append(diags, diag.Diagnostic{
+			Severity:      diag.Error,
+			Summary:       "Timeout while creating nodepool",
+			Detail:        err.Error(),
+			AttributePath: nil,
+		})
+		return
+	}
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:215]: replace with same logic as in ovh_cloud_project_kube create

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 22]
```diff
@@ -0,0 +1,405 @@
...
+			Summary:       "Failed to read nodepool",
+			Detail:        err.Error(),
+			AttributePath: nil,
+		})
+		return
+	}
+
+	d.Partial(false)
+
+	return nil
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:258]: return resourceCloudProjectKubeNodePoolRead(d, meta)

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 23]
```diff
@@ -0,0 +1,405 @@
...
+
+func resourceCloudProjectKubeNodePoolRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	projectId := d.Get("project_id").(string)
+	clusterId := d.Get("cluster_id").(string)
+
+	r := &CloudProjectKubeNodePoolResponse{}
+
+	d.Partial(true)
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:269]: to be removed

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 24]
```diff
@@ -0,0 +1,405 @@
...
+
+	endpoint := fmt.Sprintf("/cloud/project/%s/kube/%s/nodepool/%s", projectId, clusterId, d.Id())
+
+	log.Printf("[DEBUG] Will read nodepool %s from cluster %s in project %s", d.Id(), clusterId, projectId)
+
+	err := config.OVHClient.Get(endpoint, r)
+	if err != nil {
+		return fmt.Errorf("fail to get nodepool on %s: %v", endpoint, err)
+	}
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:278]: add  CheckDeleted

eg:
```
if err := config.OVHClient.Get(endpoint, r); err != nil {
       return helpers.CheckDeleted(d, err, endpoint)                                                                                                                                                                                   
  }
```

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 25]
```diff
@@ -0,0 +1,405 @@
...
+
+	err = readCloudProjectKubeNodePool(projectId, config, d, r)
+	if err != nil {
+		return fmt.Errorf("error while reading nodepool data %s: %v", d.Id(), err)
+	}
+	d.Partial(false)
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:284]: to be removed

File: ovh/resource_ovh_cloud_kubernetes_node_pool.go [hunk 26]
```diff
@@ -0,0 +1,405 @@
...
+	_ = d.Set("name", cluster.Name)
+	_ = d.Set("flavor", cluster.Flavor)
+	_ = d.Set("desired_nodes", cluster.DesiredNodes)
+	_ = d.Set("max_nodes", cluster.MaxNodes)
+	_ = d.Set("min_nodes", cluster.MinNodes)
+	_ = d.Set("monthly_billed", cluster.MonthlyBilled)
+	_ = d.Set("status", cluster.Status)
+	d.SetId(cluster.Id)
+	err = nil
+	return
```
Comment (yanndegat) [inline ovh/resource_ovh_cloud_kubernetes_node_pool.go:357]: return nil

**Review by yanndegat — CHANGES_REQUESTED** [review]
hi @d33d33, @PhilippeVienne 

thanks for this PR, this is great work.

this is probably going to be merged in the next release. there are still some stuff to be done yet. 
i'll make general comments here and more precise ones inline.

# resource naming

resources/datasources should be named after their api path
so the nodepool is should be renamed:
ovh_cloud_project_kube_node_pool -> ovh_cloud_project_kube_nodepool


# file naming
first global remark: some files should be renamed accordlingly to the resources/datasources they define.
eg: ovh/resource_ovh_cloud_kubernetes_cluster.go -> ovh/resource_ovh_cloud_project_kube.go


ovh/data_source_ovh_cloud_kubernetes_cluster.go 
ovh/data_source_ovh_cloud_kubernetes_cluster_test.go
ovh/resource_ovh_cloud_kubernetes_cluster.go
ovh/resource_ovh_cloud_kubernetes_node_pool_test.go 
ovh/resource_ovh_cloud_kubernetes_node_pool.go
ovh/resource_ovh_cloud_kubernetes_cluster_test.go 

# documentation

documentation files are missing. there should be a doc file for each datasource/resource file added
ovh/data_source_ovh_cloud_project_kube.go 
ovh/resource_ovh_cloud_project_kube.go
ovh/resource_ovh_cloud_project_kube_node_pool.go

# commits

could you squash your commits before we merge ?

File: ovh/data_source_ovh_cloud_kubernetes_cluster.go [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceCloudProjectKube() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceCloudProjectKubeRead,
+		Schema: map[string]*schema.Schema{
+			"project_id": {
```
[the remark this answers is not in the document]
Comment (pgaxatte) [reply ovh/data_source_ovh_cloud_kubernetes_cluster.go:14]: There's a typo, it's `service_name`

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#85.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#86.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#86.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#92.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#92.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#93.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#93.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#123.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#123.0 ✘  
* 
  * checks ✘

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#124.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#124.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#126.0 ✘  
* 
  * checks ✘

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#126.0 ✘  
* 
  * checks ✘

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#126.1 ✘  
* 
  * checks ✘

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#126.1 ✘  
* 
  * checks ✘

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#131.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#131.0 ✔ 
* 
  * checks ✔

**Review by yanndegat — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#167.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#167.0 ✘  
* 
  * checks ✘

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#167.1 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#169.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#169.0 ✘  
* 
  * checks ✘

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#169.1 ■

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#169.1 ■

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-check#169.0 ✔ 
* 
  * checks ✔

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#169.2 ✘  
* 
  * checks ✘

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#169.3 ✘  
* 
  * checks ✘

**Comment by ovh-cds** [comment]
CDS Report terraform-provider-ovh-testacc#169.4 ✔ 
* 
  * checks ✔

**Comment by yanndegat** [comment]
finally

thanks to all of you @PhilippeVienne @d33d33

**Pull request merged**